File: ios/Modules/Sources/Scheduled/impl/HttpScheduled.swift
import Foundation
import Net

public actor HttpScheduled: ScheduledMessages {
    private let http: HttpClient
    public init(http: HttpClient) { self.http = http }

    private struct Wrap: Codable { let items: [ScheduledMessage] }
    private struct AddBody: Codable { let room_id: Int64; let ciphertext_b64: String; let send_at: Int64 }
    private struct EditBody: Codable { let ciphertext_b64: String; let send_at: Int64? }

    public func list(roomId: Int64) async -> [ScheduledMessage] {
        let req = HttpRequest.get("/api/scheduled", query: [URLQueryItem(name: "room_id", value: "\(roomId)")])
        return (try? await http.send(req, Wrap.self).items) ?? []
    }

    public func schedule(roomId: Int64, ciphertextB64: String, sendAt: Int64) async -> ScheduledMessage? {
        guard let req = try? HttpRequest.postJson("/api/scheduled",
            body: AddBody(room_id: roomId, ciphertext_b64: ciphertextB64, send_at: sendAt)) else { return nil }
        return try? await http.send(req, ScheduledMessage.self)
    }

    public func cancel(id: Int64) async {
        let req = HttpRequest.delete("/api/scheduled/\(id)")
        _ = try? await http.send(req)
    }

    public func edit(id: Int64, newCiphertextB64: String, newSendAt: Int64?) async -> ScheduledMessage? {
        guard var req = try? HttpRequest.postJson("/api/scheduled/\(id)",
            body: EditBody(ciphertext_b64: newCiphertextB64, send_at: newSendAt)) else { return nil }
        req.method = .PATCH
        return try? await http.send(req, ScheduledMessage.self)
    }
}

File: android/app/src/main/java/sol/vortexx/android/scheduled/impl/HttpScheduled.kt
package sol.vortexx.android.scheduled.impl

import io.ktor.client.call.body
import io.ktor.client.request.*
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import kotlinx.serialization.Serializable
import sol.vortexx.android.net.impl.VortexHttpClient
import sol.vortexx.android.scheduled.api.ScheduledMessage
import sol.vortexx.android.scheduled.api.ScheduledMessages
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class HttpScheduled @Inject constructor(
    private val http: VortexHttpClient,
) : ScheduledMessages {
    @Serializable private data class Wrap(val items: List<ScheduledMessage>)
    @Serializable private data class AddBody(val room_id: Long, val ciphertext_b64: String, val send_at: Long)
    @Serializable private data class EditBody(val ciphertext_b64: String, val send_at: Long? = null)

    override suspend fun list(roomId: Long): List<ScheduledMessage> = runCatching {
        val resp = http.client.get("api/scheduled") { url { parameters.append("room_id", roomId.toString()) } }
        if (!resp.status.isSuccess()) emptyList() else resp.body<Wrap>().items
    }.getOrDefault(emptyList())

    override suspend fun schedule(roomId: Long, ciphertextB64: String, sendAt: Long): ScheduledMessage? = runCatching {
        val resp = http.client.post("api/scheduled") {
            contentType(ContentType.Application.Json)
            setBody(AddBody(roomId, ciphertextB64, sendAt))
        }
        if (!resp.status.isSuccess()) null else resp.body<ScheduledMessage>()
    }.getOrNull()

    override suspend fun cancel(id: Long) {
        runCatching { http.client.delete("api/scheduled/$id") }
    }

    override suspend fun edit(id: Long, newCiphertextB64: String, newSendAt: Long?): ScheduledMessage? = runCatching {
        val resp = http.client.patch("api/scheduled/$id") {
            contentType(ContentType.Application.Json)
            setBody(EditBody(newCiphertextB64, newSendAt))
        }
        if (!resp.status.isSuccess()) null else resp.body<ScheduledMessage>()
    }.getOrNull()
}
